Config option to allow custom parameter syntax

not-night-but opened the pull request:
This follows the same syntax that is allowed by sql-formatter, which we use in beekeeper-studio

```ts
interface ParamTypes {
  positional?: boolean, // ansi standard positional '?' parameters
  numbered?: Array<"?" | ":" | "$">, // psql like numeric parametrs '$1'
  named?: Array<":" | "@" | "$">, // mssql like named parameters ':name'
  quoted?: Array<":" | "@" | "$">, // quoted parameters, like bigquery ':`name here`'
  custom?: Array<string> // regex describing a custom parameter, eg. '/\{[a-zA-Z0-9_]+\}/' allows '{name}' as a param
}
```

MasterOdin commented:
Would this negate the need for #77 or does beekeeper want both?

I'm assuming the need here is that beekeeper wants to allow people to specify variables in any format or something regardless of what the database itself usually uses for parameters?

MasterOdin commented on test/tokenizer/index.spec.ts:
```diff
@@ -376,6 +376,262 @@ describe('scan', () => {
+              end: 1,
+            },
+            {
+              type: 'parameter',
+              value: 'unknown',
```
This doesn't seem right to me, where if it's not a parameter as it's not a number in this case, we shouldn't be classifying it as such.

not-night-but replied:
How do we want to classify it? I believe this was in the old code

MasterOdin replied:
I think it would make more sense to be `type: 'unknown'` if we don't think it's a variable because the given dialect does not support that variable format, and the value should _always_ be the actual token value.

not-night-but commented:
@MasterOdin I believe it would yes. I actually didn't see that PR till I filed this one (oops), but the idea here was to have the same config schema as sql-formatter, so that objects could be passed to both (as sql-formatter can replace the params). This also provides more options

MasterOdin commented on test/tokenizer/index.spec.ts:
```diff
@@ -376,6 +376,262 @@ describe('scan', () => {
       });
+
+      describe('custom parameters', () => {
+        it('should allow positional parameters for all dialects', () => {
```
For all the tests here, should move the foreach over dialects (and any inner foreachs) up out of the test, where each dialect is its own testcase, where it's useful to know if one particular dialect was broken or all of them with a given change.

MasterOdin commented on test/index.spec.ts:
```diff
@@ -22,6 +23,29 @@ describe('identify', () => {
     ]);
   });
 
+  it('should identify custom parameters', () => {
```
Can we have a test to validate that setting `paramTypes` will override the default parameter matching for a given adapter, e.g. if I have `positional: true` set, then numbered params will be ignored, or whatever the behavior is with `sql-formatter` in dealing with this.

MasterOdin commented on src/index.ts:
```diff
@@ -21,7 +21,8 @@ export function identify(query: string, options: IdentifyOptions = {}): Identify
   }
 
-  const result = parse(query, isStrict, dialect, options.identifyTables);
+  const result = parse(query, isStrict, dialect, options.identifyTables, options.paramTypes);
```
I think that instead of passing `options.paramTypes` undefined through if not set, we should set default values that are adapter specific, so that all adapters go through the same logic of parsing parameters, vs having both the `scanCustomParameter` function as well as adapter specific handlers below.

So like:
```
if (!options.paramTypes) {
  if (dialect === 'sqlite') {
    options.paramTypes = { positional: true };
  }
  elseif (dialect === 'psql') {
    options.paramTypes = { numbered: [ '$' ] };
  }
  // other options...
}
```

We could also do this transformation in the `tokenizer` as well, I've no particularly strong preference, just think we should do this instead of having both code paths.

not-night-but replied:
Great idea, will do

not-night-but commented:
@MasterOdin @rathboma 

I've implemented the requested changes, minus how we classify params that fail to be scanned in `scanParameter` as I'm not sure how we want to handle it. I just left it as how we handled it before, but I'm open to changing that as well. Though I'm not sure how much that branch is even hit, as `isParameter` *should* prevent that from happening.

A couple things:
1. I'm not sure `defaultParamTypesFor` should be where it is, feels like a utils fn, I just didn't want to add a file just for the function
2. In that function, I kept the default behaviour for each dialect, and added bigquery, but sql-formatter supports a couple more by default for some dialects (eg sqlite has 4 different syntaxes for params). I'm just hesitant to change default behaviour and possibly break applications using identifier. Thoughts?

rathboma commented:
@not-night-but I think we should adhere to specs - so if SQLite has 4 different supported param types, that is what we should show. We can always increment the version if we feel like it's a bigger break

MasterOdin commented:
Yeah, making sqlite adhere more closely to the spec, and just returning more correctly for variable identification seems like a minor release, especially since supporting positional variables shouldn't change, would only affect people using a mix of variable types, but feels like we're incorrect in that instance.

not-night-but commented:
@MasterOdin @rathboma Implemented the changes:

1. `scanParameter` now returns an unknown token if it can't determine what type the param is
2. Sqlite now supports positional, `?1` numbered, `:named`, and `@named` parameters. As a result, one test was removed because it's no longer a possible case

MasterOdin commented on src/tokenizer.ts:
```diff
@@ -253,52 +257,97 @@ function scanString(state: State, endToken: Char): Token {
-      nextChar = read(state);
-    } while (nextChar !== null && !isNaN(Number(nextChar)) && !isWhitespace(nextChar));
+function scanParameter(state: State, dialect: Dialect, paramTypes: ParamTypes): Token {
+  const curCh: any = state.input[state.start];
```
Can we give this a proper type?

not-night-but commented on src/tokenizer.ts:
```diff
@@ -413,18 +462,36 @@ function isString(ch: Char, dialect: Dialect): boolean {
+}
+
+function isParameter(ch: Char, state: State, paramTypes: ParamTypes): boolean {
+  const curCh: any = ch;
```
This is just to get around typescript misbehaving so we don't have to cast everywhere

MasterOdin replied:
Yeah, you cannot use `.includes` of two different types (even if one is just a subset of strings), and have to use `.some` (see 99870c408095cd5b99ed730eb8be385f5f72683f for example). Not really "typescript misbehaving" so much as just a limitation of type systems, and not a limitation that justifies reducing type safety via usage of `any`.

MasterOdin commented on src/defines.ts:
```diff
@@ -76,10 +76,20 @@ export type StatementType =
+  positional?: boolean;
+  numbered?: ('?' | ':' | '$')[];
+  named?: (':' | '@' | '$')[];
+  quoted?: (':' | '@' | '$')[];
```
Confirming that there is not any error checking if someone passes a value that's not in this array of values? I'm fine merging this as is, and then making my own PR doing this.

not-night-but replied:
Correct. There is no error checking besides typescript type checking
